backend/ping.py

```diff
@@ -0,0 +1,178 @@
+    try:
+        sock = socket.socket(address_family, socket_type, protocol)
+    except socket.error:
+        return None, False
+    sock_fd = sock.fileno()
+    fd_flags = fcntl.fcntl(sock_fd, fcntl.F_GETFL)
+    fd_flags |= os.O_NONBLOCK
+    fcntl.fcntl(sock_fd, fcntl.F_SETFL, fd_flags)
+    try:
+        sock.connect(address_tuple)
+    except socket.error as e:
+        if e.errno == errno.EINPROGRESS:
+            return sock, False
+        else:
+            sock.close()
+            return None, False
+    else:
+        sock.close()
+        return None, True
+
+
+def _create_socket(struct_getaddrinfo):
+    return _create_socket_2(struct_getaddrinfo[0],
+                              struct_getaddrinfo[1],
+                              struct_getaddrinfo[2],
+                              struct_getaddrinfo[4])
+
+
+def _wait_for_any_change(socket_fds):
+    select_results = select.select(socket_fds,
+                                   socket_fds,
+                                   socket_fds,
+                                   None)
+    return list(set(itertools.chain.from_iterable(select_results)))
+
+
+def _check_connection(sock):
+    try:
+        sock.getpeername()
+    except socket.error as e:
+        if e.errno != errno.ENOTCONN:
+            raise
+        result = False
+    else:
+        result = True
+    finally:
+        sock.close()
+    return result
+
+
+def _scan(endpoints):
+    results = {}
+    sockets = {}
+    try:
+        for host, port in endpoints:
+            getaddrinfo_results = socket.getaddrinfo(host,
+                                                     port,
+                                                     socket.AF_UNSPEC,
+                                                     socket.SOCK_STREAM)
+            for r in getaddrinfo_results:
+                sock, is_open = _create_socket(r)
+                if is_open:
+                    results[host, port] = True
+                    break
+                elif sock:
+                    # Yes, a nested tuple.
+                    sockets[sock.fileno()] = sock, (host, port)
+                    break
+            # If not broken...
+            else:
+                results[host, port] = False
+        while sockets:
+            fds = _wait_for_any_change(sockets.keys())
+            for fd in fds:
+                sock, host_and_port = sockets[fd]
+                results[host_and_port] = _check_connection(sock)
+                del sockets[fd]
+    finally:
+        for s, _ in sockets.itervalues():
+            s.close()
+        sockets.clear()
+    return results
+
+
+def _split(endpoints):
+    '''A generator yielding parts of the given endpoint list split
+    in such way, that soft maximum open file descriptor resource
+    limit **should** not be exceeded.'''
+    i = 0
+    while i < len(endpoints):
+        limit = _get_file_descriptor_resource_limit()
```

Comment:
Theoretically it can change any moment. That is why I've initially put it here. But later I've had the same thought as You. What is Your opinion on this? Maybe move it and add a comment, since it's not synchronised anyway?